src/compiler/checker.ts

```diff
@@ -29172,8 +29172,21 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         }
 
         function narrowTypeByBooleanComparison(type: Type, expr: Expression, bool: BooleanLiteral, operator: BinaryOperator, assumeTrue: boolean): Type {
-            assumeTrue = (assumeTrue !== (bool.kind === SyntaxKind.TrueKeyword)) !== (operator !== SyntaxKind.ExclamationEqualsEqualsToken && operator !== SyntaxKind.ExclamationEqualsToken);
-            return narrowType(type, expr, assumeTrue);
+            const isTrueLiteral = bool.kind === SyntaxKind.TrueKeyword;
+            const equalsOperator = operator === SyntaxKind.EqualsEqualsToken || operator === SyntaxKind.EqualsEqualsEqualsToken;
+            // pass down `assumeTrue` computed using boolean xors
+            const narrowedType = narrowType(type, expr, (assumeTrue !== isTrueLiteral) !== equalsOperator);
+            if (!assumeTrue && type.flags & TypeFlags.Union && optionalChainContainsReference(expr, reference)) {
+                const types = [narrowedType];
+                if (containsUndefinedType(type)) {
+                    types.push(undefinedType);
+                }
+                if (containsType((type as UnionType).types, nullType)) {
+                    types.push(nullType);
+                }
+                return getUnionType(types);
+            }

```

Comment:
I'm still thinking if this is the right place to "add back" those nullable types but in the meantime an extended test suite run could be insightful (however, this likely is not a popular pattern so maybe chances of it catching any potential issues is pretty low).

It feels like `narrowTypeByCallExpression` could be a better place for this... but it's much easier to do it here. `narrowTypeByCallExpression` receives a "flipped" `assumeTrue` so it isn't aware that it should not filter out null/undefined in what is the "true" `!assumeTrue` branch because it sees `assumeTrue === true`.

Reply:
note for myself: `optionalChainContainsReference ` is too broad of a check, I should add tests with optional chains and identifier predicates